### PR TITLE
VRF and Gameplay contract update

### PR DESCRIPTION
**Pull request by AceVikings**
This update contains contract for Random number generation and token contract update for gameplay functionalities.

### CODE REVIEW

**Comment by Niraj-Kamdar** [comment]
Do discussed changes: 
createRoom - mapping random to room id , call request random for next room
getRandom - take room id and send mapped random back 
reward - playerId - token address, get owner from it and give them allowance
redeem reward - set allowance 0 and transfer it to user

File: contracts/Accounts.sol [hunk 1]
```diff
@@ -2,15 +2,26 @@
...
+    }
     
     function signUp(string memory accountURI, uint256 accountID) external{
         _safeMint(_msgSender(), accountID);
         _setTokenURI(accountID, accountURI);
+        TI.signUpReward(msg.sender, 100);
+    }
+    
+    function AddTIAddress(address tokenAddress) public {
+        TI = TokenInterface(tokenAddress);
```
Comment (Niraj-Kamdar) [inline contracts/Accounts.sol]: This shouldn't be public.

File: contracts/CryptoMafiaCoin.sol [hunk 2]
```diff
@@ -4,12 +4,45 @@ pragma solidity ^0.8.0;
...
+    
+    address addressRandomGenerator;
+    
+    address contractAddress;
+    
+    RandomInterface RNG;
+    
+    AccountInterface AI;
+    
+    mapping(uint256=>uint256) roomStake;
```
Comment (Niraj-Kamdar) [inline contracts/CryptoMafiaCoin.sol]: What's the need for this?

File: contracts/CryptoMafiaCoin.sol [hunk 3]
```diff
@@ -4,12 +4,45 @@ pragma solidity ^0.8.0;
...
+    mapping(uint256=>mapping(uint256=>address)) playerID;
```
Comment (Niraj-Kamdar) [inline contracts/CryptoMafiaCoin.sol]: Still double mapping?

File: contracts/CryptoMafiaCoin.sol [hunk 4]
```diff
@@ -4,12 +4,45 @@ pragma solidity ^0.8.0;
...
+    mapping(uint256=>uint256) roomRandomNumber;
+    mapping(uint256=>bool) randomNumIsTaken;
```
Comment (Niraj-Kamdar) [inline contracts/CryptoMafiaCoin.sol]: No need. if first mapping is returning 0 then random number haven't fullfilled

File: contracts/CryptoMafiaCoin.sol [hunk 5]
```diff
@@ -4,12 +4,45 @@ pragma solidity ^0.8.0;
...
+    }
+    
+    
+    function setAccountAddress(address accountAddress) public {
+        AI = AccountInterface(accountAddress);
+    }
+    
+    function setRandomAddress(address accountAddress) public {
+        RNG = RandomInterface(accountAddress);
     }
```
Comment (Niraj-Kamdar) [inline contracts/CryptoMafiaCoin.sol]: Imagine a random person changing the addess and causing contract to misbehave. In worst case hack the contract and drain the fund

File: contracts/RandomGenerator.sol [hunk 6]
```diff
@@ -0,0 +1,46 @@
+pragma solidity ^0.6.0;
+
+import "https://github.com/smartcontractkit/chainlink/blob/master/evm-contracts/src/v0.6/VRFConsumerBase.sol";
+
+contract RandomGenerator is VRFConsumerBase  {
```
Comment (Niraj-Kamdar) [inline contracts/RandomGenerator.sol:7]: Maintain roomId => randomNo mapping and requestId => roomid mapping here.

**Review by Niraj-Kamdar — CHANGES_REQUESTED** [review]
Fix the suggested issues and maintaining roomId=> randomNo mapping is must for correct behaviour.

File: contracts/Accounts.sol (hunk 1 above)
Comment (AceVikings) [reply contracts/Accounts.sol]: It has to be public but I'll change it to onlyOwner. We need to set the contract address after all the contracts have been deployed therefore this.

File: contracts/CryptoMafiaCoin.sol (hunk 2 above)
Comment (AceVikings) [reply contracts/CryptoMafiaCoin.sol]: This tells us the total amount staked in the room for a given roomId

File: contracts/CryptoMafiaCoin.sol (hunk 4 above)
Comment (AceVikings) [reply contracts/CryptoMafiaCoin.sol]: If we check key value mapping for just if a value exists for the key that doesn't guarantee that the value would be unique, basically I am check if a random number is already in use or not

File: contracts/CryptoMafiaCoin.sol (hunk 5 above)
Comment (AceVikings) [reply contracts/CryptoMafiaCoin.sol]: I'll change it to onlyOwner

File: contracts/RandomGenerator.sol (hunk 6 above)
Comment (AceVikings) [reply contracts/RandomGenerator.sol:7]: Do we need requestId mapping ?

Comment (AceVikings) [reply contracts/RandomGenerator.sol:7]: roomId=>randomNo mapping already exists in GameCoin contract where its being used, any specific reason to move that here ?

Comment (Niraj-Kamdar) [reply contracts/RandomGenerator.sol:7]: Yes, if you don't there will be dirty update problem.

File: contracts/CryptoMafiaCoin.sol (hunk 4 above)
Comment (Niraj-Kamdar) [reply contracts/CryptoMafiaCoin.sol]: We don't need unique random no. we just need a random no. for roomId. It's okay if sometime 2 random no. are equal

File: contracts/CryptoMafiaCoin.sol [hunk 7]
```diff
@@ -24,4 +56,58 @@ contract CryptoMafiaCoin is ERC20, Ownable {
     function mint(address receiver, uint amount) onlyAccountsHolder public {
         _mint(receiver, amount);
     }
+    
+    
+    function RequestRandomNumberCMC(uint256 seed) public returns(bytes32){
```
Comment (Niraj-Kamdar) [inline contracts/CryptoMafiaCoin.sol]: use a nonce as a seed.

File: contracts/CryptoMafiaCoin.sol [hunk 8]
```diff
@@ -24,4 +56,58 @@ contract CryptoMafiaCoin is ERC20, Ownable {
...
+    }
+    
+    function getRandom(uint256 roomId) public view returns(uint256){
+        return roomRandomNumber[roomId];
+    }
+    
+    function joinRoom(uint256 roomId) public payable
+    {   
+        transfer(contractAddress,10);
+        roomStake[roomId] = roomStake[roomId] + 10;
```
Comment (Niraj-Kamdar) [inline contracts/CryptoMafiaCoin.sol]: why maintain roomStake? it will be constant 8*fee

**Review by Niraj-Kamdar — CHANGES_REQUESTED** [review]

File: contracts/CryptoMafiaCoin.sol [hunk 9]
```diff
@@ -24,4 +53,43 @@ contract CryptoMafiaCoin is ERC20, Ownable {
...
+            increaseAllowance(AI.ownerOf(playerId[j]),disbursalAmount);
+      }
+      
+     
+    }
+    
+    
+    function signUpReward(address userAddress,uint256 amount) public {
+        _mint(userAddress,amount);
+    }
```
Comment (Niraj-Kamdar) [inline contracts/CryptoMafiaCoin.sol]: use mint function directly

File: contracts/CryptoMafiaCoin.sol [hunk 10]
```diff
@@ -24,4 +53,43 @@ contract CryptoMafiaCoin is ERC20, Ownable {
     function mint(address receiver, uint amount) onlyAccountsHolder public {
         _mint(receiver, amount);
     }
```
Comment (Niraj-Kamdar) [inline contracts/CryptoMafiaCoin.sol:55]: this one directly in ERC721 with interface

**Review by Niraj-Kamdar — CHANGES_REQUESTED** [review]

File: contracts/RandomGenerator.sol [hunk 11]
```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.0;
+
+import "https://github.com/smartcontractkit/chainlink/blob/master/evm-contracts/src/v0.6/VRFConsumerBase.sol";
```
Comment (Niraj-Kamdar) [inline contracts/RandomGenerator.sol]: use this github URL isn't reliable, import "@chainlink/contracts/src/v0.6/VRFConsumerBase.sol";

**Pull request merged**